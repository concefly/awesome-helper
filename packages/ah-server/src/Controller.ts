import { Schema } from 'jsonschema';
import { IContext } from '.';
import { Service } from './Service';

type IRouterMethod = 'GET' | 'POST' | 'DELETE' | 'PUT';

export type IControllerMapperItem = {
  path: string;
  method: IRouterMethod | IRouterMethod[];
  handler: (ctx: IContext, query?: any) => Promise<any>;
  query?: { schema: Schema };
};

export abstract class Controller extends Service {
  abstract readonly mapper: IControllerMapperItem[];
}
