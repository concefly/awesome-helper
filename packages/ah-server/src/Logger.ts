import dayjs from 'dayjs';

export class Logger {
  private name = 'APP';

  setName(name: string) {
    this.name = name;
    return this;
  }

  private ft(level: string, msg: string) {
    const timestamp = dayjs().format('YYYY-MM-DD HH:mm:ss.SSS');
    return `[${timestamp}|${this.name}|${level}] ${msg}`;
  }

  extend(subName: string): Logger {
    const newName = [this.name, subName].join('.');
    return new Logger().setName(newName);
  }

  info(msg: string) {
    console.info(this.ft('INFO', msg));
  }

  error(msg: string) {
    console.error(this.ft('ERROR', msg));
  }
}
