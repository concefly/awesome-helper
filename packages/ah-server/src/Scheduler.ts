import { Service } from './Service';

/** 基础调度服务 */
export class Scheduler extends Service {
  cron?: string;
  interval?: number;

  async invoke(): Promise<void> {}
}
