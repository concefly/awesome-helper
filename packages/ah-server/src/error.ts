// eslint-disable-next-line
export enum ErrorTypeEnum {
  custom = 'custom',
}

const createBaseBizError = (
  message: string,
  type: ErrorTypeEnum,
  status: number,
  code: string = 'unknown'
) => {
  return Object.assign(new Error(message), { type, status, code });
};

/** 业务错误，和系统错误区分开 */
export const createBizError = (message: string, code?: string) => {
  return createBaseBizError(message, ErrorTypeEnum.custom, 400, code);
};

export const createUnAuthError = (message: string, code?: string) => {
  return createBaseBizError(message, ErrorTypeEnum.custom, 401, code);
};

export const createForbiddenError = (message: string, code?: string) => {
  return createBaseBizError(message, ErrorTypeEnum.custom, 403, code);
};
