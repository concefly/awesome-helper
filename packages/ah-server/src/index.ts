import 'koa';
import { DefaultContext, DefaultState, ParameterizedContext } from 'koa';

export interface IApplication {}
export interface IService {}

export interface IContext extends ParameterizedContext<DefaultState, DefaultContext> {}
export interface IConfig {
  LOCAL_PORT: number;
}

// fix koa plugin type
declare module 'koa' {
  interface Request {
    // koaBody type
    body?: any;
  }
}

export * from './App';
export * from './Config';
export * from './Logger';
export * from './Controller';
export * from './Service';
export * from './util';
export * from './Scheduler';
export * from './error';
