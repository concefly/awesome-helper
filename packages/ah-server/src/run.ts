import { App, Service, IService, Config } from '.';

export class AService extends Service {}
export class BService extends Service {}

class DemoApp extends App {
  service: IService = {
    a: new AService(this),
    b: new BService(this),
  };
}

(async () => {
  const app = new DemoApp(new Config());
  await app.start();
})().catch(e => console.error(e));
