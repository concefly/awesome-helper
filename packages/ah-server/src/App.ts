import { CronJob } from 'cron';
import Koa from 'koa';
import Router from 'koa-router';
import koaBody from 'koa-body';
import urllib, { RequestOptions } from 'urllib';
import { Logger } from './Logger';
import { Scheduler } from './Scheduler';
import { Service } from './Service';
import { IConfig, IContext, IService } from '.';
import { ErrorTypeEnum } from './error';
import { pick, tryParseIntProperty, validate } from './util';
import { Controller } from './Controller';
import { Server } from 'http';

declare module '.' {
  interface IApplication extends App {}

  // eslint-disable-next-line
  interface IContext {
    validate: typeof validate;
    app: IApplication;
  }
}

export abstract class App extends Koa {
  constructor(readonly config: IConfig) {
    super();
  }

  // 注入 app 扩展
  public abstract service: IService = {};
  /** controller 列表 */
  public abstract controllerList: Controller[] = [];
  /** 定时任务列表 */
  public abstract schedulerList: Scheduler[] = [];

  public logger = new Logger();

  private server?: Server;

  public async curl<T>(url: string, opt?: RequestOptions) {
    return urllib.request<T>(url, opt);
  }

  private async initCommon() {
    // 扩展 ctx
    Object.assign(this.context, { validate, app: this });

    // 全局错误
    this.on('error', err => {
      let msg = err.message || err;
      if (err.stack) msg += '\n' + err.stack;

      this.logger.error(msg);
    });
  }

  private async initService() {
    const list: Service[] = Object.values(this.service);
    await Promise.all(list.map(s => s.init?.()));

    this.logger.info(`service: ${list.map(s => s.name)}`);
  }

  private async initController() {
    this.use(koaBody());

    // 构造 router
    const router = new Router<any, IContext>();
    this.controllerList.forEach(ctrlIns => {
      ctrlIns.mapper.forEach(m => {
        this.logger.info(
          `register controller: ${m.method} ${m.path} -> ${ctrlIns.name}.${m.handler.name}`
        );

        router.register(m.path, Array.isArray(m.method) ? m.method : [m.method], async ctx => {
          try {
            const q = m.query
              ? ctx.validate<any>(
                  {
                    ...tryParseIntProperty(ctx.params),
                    ...tryParseIntProperty(ctx.request.query),
                    ...ctx.request.body,
                  },
                  m.query.schema
                )
              : undefined;

            const data = await m.handler.call(ctrlIns, ctx, q);

            if (data) {
              ctx.set('content-type', 'application/json');
              ctx.body = { ...ctx.body, data };
            }
          } catch (err) {
            // 自定义异常
            if (Object.values(ErrorTypeEnum).includes(err.type)) {
              ctx.status = err.status;
              ctx.body = pick(err, ['message', 'type', 'code', 'status']);
              return;
            }

            // 其他异常外抛
            throw err;
          }
        });
      });
    });

    this.use(router.routes());
    this.use(router.allowedMethods());

    await Promise.all(this.controllerList.map(c => c.init?.()));
  }

  /** 启动定时调度 */
  private async initScheduler() {
    const list = this.schedulerList;
    if (list.length === 0) return;

    const schedulerLogger = this.logger.extend('Scheduler');
    const startList = list.map(s => {
      if (s.cron) {
        return () => {
          new CronJob(
            s.cron!,
            () => {
              s.invoke().catch(e => {
                schedulerLogger.error(`${s.name} error: ${e.message || e}`);
              });
            },
            undefined,
            false,
            undefined,
            undefined,
            true
          ).start();
        };
      } else if (s.interval) {
        return () => {
          const invoke = () => {
            s.invoke()
              .then(() => setTimeout(invoke, s.interval))
              .catch(e => {
                schedulerLogger.error(`${s.name} error: ${e.message || e}`);
                setTimeout(invoke, s.interval);
              });
          };

          invoke();
        };
      }
    });

    startList.forEach(s => s?.());

    this.logger.info(`start scheduler: ${list.map(s => s.name)}`);
  }

  public async start() {
    this.logger.info(this.config.sequelize());

    await this.initCommon();
    await this.initService();
    await this.initController();
    await this.initScheduler();

    const port = this.config.LOCAL_PORT;
    this.server = this.listen(port);
    this.logger.info(`app start at localhost:${port}`);
  }

  public async stop(): Promise<void> {
    if (!this.server) return;
    const server = this.server;

    return new Promise((resolve, reject) => {
      // 优雅退出
      // @see https://zhuanlan.zhihu.com/p/275312155?utm_source=wechat_session&utm_medium=social&utm_oi=39191756931072
      server.close(err => {
        if (err) return reject(err);
        return resolve();
      });
    });
  }
}
