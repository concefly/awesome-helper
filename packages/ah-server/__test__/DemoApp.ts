import { App, Controller, IContext, IControllerMapperItem, IService, Service } from '../src';

class EchoService extends Service {
  echo(text: string) {
    return text;
  }
}

class EchoController extends Controller {
  mapper: IControllerMapperItem[] = [
    {
      path: '/echo',
      method: ['GET', 'POST'],
      handler: this.echo,
      query: {
        schema: {
          type: 'object',
          properties: { text: { type: 'string' } },
        },
      },
    },
  ];

  async echo(_ctx: IContext, q: { text?: string }) {
    const output = (this.service as any).echo.echo(q.text);
    return output;
  }
}

export class TestApp extends App {
  service: IService = {
    echo: new EchoService(this),
  };
  controllerList: Controller[] = [new EchoController(this)];
  schedulerList = [];
  logger = {
    info() {},
    error(msg: string) {
      throw new Error(msg);
    },
  } as any;
}
