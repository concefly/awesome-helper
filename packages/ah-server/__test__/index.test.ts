import { HttpClientResponse } from 'urllib';
import { Config, pick } from '../src';
import { TestApp } from './DemoApp';

function inspectResp(resp: HttpClientResponse<any>) {
  return pick(resp, ['data', 'status']);
}

describe('App', () => {
  const config = new Config();
  const app = new TestApp(config);

  beforeAll(() => app.start());
  afterAll(() => app.stop());

  it('curl', async () => {
    const r1 = await app.curl<any>('http://localhost:10001/echo?text=hi', { dataType: 'json' });
    expect(inspectResp(r1)).toMatchSnapshot();
  });

  it('post body', async () => {
    const r2 = await app.curl<any>('http://localhost:10001/echo', {
      method: 'POST',
      contentType: 'json',
      dataType: 'json',
      data: {
        text: 'hi',
      },
    });
    expect(inspectResp(r2)).toMatchSnapshot();
  });

  it('validate 400', async () => {
    const r3 = await app.curl<any>('http://localhost:10001/echo?text=100', { dataType: 'json' });
    expect(inspectResp(r3)).toMatchSnapshot();
  });
});
