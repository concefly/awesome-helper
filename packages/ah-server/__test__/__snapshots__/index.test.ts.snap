// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App curl 1`] = `
Object {
  "data": Object {
    "data": "hi",
  },
  "status": 200,
}
`;

exports[`App post body 1`] = `
Object {
  "data": Object {
    "data": "hi",
  },
  "status": 200,
}
`;

exports[`App validate 400 1`] = `
Object {
  "data": Object {
    "code": "unknown",
    "message": "is not of a type(s) string",
    "status": 400,
    "type": "custom",
  },
  "status": 400,
}
`;
